design/ux/two_phase_game_move_execution.md

```diff
@@ -0,0 +1,69 @@
+
+### Game pattern - Decouple move recording from move execution
+
+When the user interacts with a game, they just commit to a move which the game stores in an internal queue. No processing should happen in this first phase.
+The execution will be performed in phase 2 - as a separate transaction.
+
+Note: The "execution" transaction can't be triggered by the user because then they can perform the proxy attack by executing both phases inside the proxy.
+
+To implement this pattern securely, there must be a trusted third party that calls "execute" on behalf of users. 
+This solution is not ideal because it affects the UX - the user will have to wait at least 2 blocks for a response, and it introduces a third party. 
+There is also the issue of the execution cost. The third party has to be somehow paid for this service.
+
+
+### Platform support - execution callback
+
+TEN already has a system contract that is executed at the end of every block. The execution of the sys contract is triggered by the platform as a "synthetic" transaction.
+
+If games had a way to register themselves with this system contract, then it could call the execution phase automatically at the end of each batch.
+
+This solves the UX problem - because the user would get the response at the same time, and it also removes the need for a trusted third party.
+
+#### Implementation 
+
+Ideally, the registration of the callback would be a "smart contract"-only action.  
+Basically, the system contract that is called automatically will have a function where it allows anyone to register a callback.
+
+
+#### Paying for the execution
+
+This is the trickiest component of the solution.
+
+On a high level, the most likely requirement is that the Game would ask the users to pay an amount that covers both the recording of the move and the execution of an average move.
+As part of the "phase 1" transaction, the money for the move recording goes to the platform, while the estimated money for "phase 2" goes to an account owned by the game.
+
+Assuming we implement the callback as described in the previous section, we need a way from solidity to calculate the gas cost of executing each move. 
+This appears to be possible using a workaround with the `gasLeft()` function.
+
+##### Possible Implementation of execution payment
+
+1. The game developer has to estimate how much an average move will cost to execute and add some margin.
+2. During the "phase 1" transaction, the game dev will transfer some `value` from the caller to an address controlled by the sys contract.
+   This address was allocated to this game during the callback registration (similar to how you have to reference a payment when you move money to some broker).
+   Note: To the user, it will be equivalent to gas spent for the transaction.
+4. At the end of the batch, the sys contract will invoke the "phase 2" function of the game - one move at a time.
```

Comment:
I think I'll say pretty much what's in the PR already, but just to be clear - system contract will invoke registered callbacks in order of registration; so we can think of it as a queue. registerCallback pushes in it, and then we pop. So if there is some ordering where contract A registers callback and then calls contract B in same transaction which registers another, then it comes back and third one is registered it should all be executed in order. Not sure if this will ever be of any practical use, but I think this is the intuitive expectation

Reply:
umm. We need to think about this a bit more.